### PR TITLE
Bug 1273096 - Backfill jobs directly against Taskcluster API

### PR DESCRIPTION
**Pull request by imbstack**
This PR contains a few changes:

1. tc credentials are stored in local storage upon login
2. a thTaskcluster service exists to use those credentials via a vendored taskcluster client
3. the backfill action will check to see if a job is a taskcluster job, and if so, it uses the taskcluster client to submit a task directly to taskcluster without going through pulse_actions.

*This is not quite finished yet.* In particular:

1. I don't have a plan for logging the user out when taskcluster credentials are no longer valid (every 3 days at least).
2. I expect a good portion of this code is in the wrong place and needs to be moved. I'm not really familiar with angular; I stuck it where it was easiest for now and am happy to do whatever refactoring is needed.

With that said, this does pretty much work, afaict. I tried this out on my local treeherder and it got all the way through to submitting a job to tc before tc yelled about me lacking scopes and that is correct behavior considering that the user I was using had incomplete scopes. I'll start pulling down new data for the autoland branch asap so that I can test for real without disturbing anything.*

Please critique my angular-foo and let me know where to move things properly. Also I would like to figure out a plan for when tc creds expire before the th login does.

* We've been using certain platforms on autoland to test backfilling so far.

---
This change is [Reviewable](https://reviewable.io/reviews/mozilla/treeherder/2057)

### CODE REVIEW

File: ui/js/components/auth.js [hunk 1]
```diff
@@ -161,6 +162,14 @@ treeherder.component("loginCallback", {
                 key: $location.search().accessToken,
                 algorithm: 'sha256'
             };
+
+            // Cribbed from taskcluster-tools. Save this for interacting with tc
+            const results = $location.search();
+            if (typeof results.certificate === 'string') {
```
Comment (edmorley) [inline ui/js/components/auth.js]: I think we can use `certificate` from above rather than needing to use $location again? :-)

**Comment by imbstack** [comment]
I now believe this to be ready for merging (plus-or-minus review comments) _after_ [bug 1278986](https://bugzilla.mozilla.org/show_bug.cgi?id=1278986) has been resolved. If that bug is going to take a while to sort out, then I think we have to live with pulse_actions having powerful scopes for now and wrap all of this tc logic in a try-catch that sends a pulse message on failure. If we can resolve that bug soon, then this just needs to go through a few rounds of review and we can ship-it.

**Comment by imbstack** [comment]
@camd The bug for permissions updates has been resolved, so as soon as we can get this reviewed, it's ready to ship!

**Comment by imbstack** [comment]
Review status: 0 of 9 files reviewed at latest revision, 5 unresolved discussions.

---

*[ui/js/models/resultsets_store.js, line 340 at r3](https://reviewable.io:443/reviews/mozilla/treeherder/2057#-K__g3iqlVvZniU8vu6m-r3-340:-KasaIsJSSovMAbNnBjx:bedg9j) ([raw file](https://github.com/mozilla/treeherder/blob/9cf1e5c238d2c568700c3ad928b40ebd9c47a3f8/ui/js/models/resultsets_store.js#L340)):*
<details><summary><i>Previously, camd wrote…</i></summary><blockquote>

can this function replace ``getGeckoDecisionTaskId``?
</blockquote></details>

I'm pretty sure it can replace it with a bit of work. The only difference is that this is an async function and `getGeckoDecisionTaskId` is sync. However, `getGecko...` isn't generally that useful unless `fetchGecko...` has happened first, so it makes sense to me to do the refactor. I'll make that change.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mozilla/treeherder/2057)*

**Comment by imbstack** [comment]
Review status: 0 of 11 files reviewed at latest revision, 5 unresolved discussions.

---

*[ui/js/components/auth.js, line 168 at r1](https://reviewable.io:443/reviews/mozilla/treeherder/2057#-K__g3iqlVvZniU8vu6l-r1-168:-Kat9x1nvgwzQNxzd9p7:b-896fix) ([raw file](https://github.com/mozilla/treeherder/blob/caef0117e36e638b281d28ab76d7e82f3cee940b/ui/js/components/auth.js#L168)):*
<details><summary><i>Previously, edmorley (Ed Morley) wrote…</i></summary><blockquote>

I think we can use `certificate` from above rather than needing to use $location again? :-)
</blockquote></details>

Done.

---

*[ui/js/models/resultsets_store.js, line 340 at r3](https://reviewable.io:443/reviews/mozilla/treeherder/2057#-K__g3iqlVvZniU8vu6m-r3-340:-Kat9v5VYhIRgKC8jnEZ:b-896fix) ([raw file](https://github.com/mozilla/treeherder/blob/9cf1e5c238d2c568700c3ad928b40ebd9c47a3f8/ui/js/models/resultsets_store.js#L340)):*
<details><summary><i>Previously, imbstack (Brian Stack) wrote…</i></summary><blockquote>

I'm pretty sure it can replace it with a bit of work. The only difference is that this is an async function and `getGeckoDecisionTaskId` is sync. However, `getGecko...` isn't generally that useful unless `fetchGecko...` has happened first, so it makes sense to me to do the refactor. I'll make that change.
</blockquote></details>

Done.

---

*[ui/plugins/controller.js, line 49 at r3](https://reviewable.io:443/reviews/mozilla/treeherder/2057#-K__g3iqlVvZniU8vu6n-r3-49:-KatB6Lzo2YPqiPt_TAZ:b-896fix) ([raw file](https://github.com/mozilla/treeherder/blob/9cf1e5c238d2c568700c3ad928b40ebd9c47a3f8/ui/plugins/controller.js#L49)):*
<details><summary><i>Previously, camd wrote…</i></summary><blockquote>

Would you document here why this is necessary?
</blockquote></details>

Done.

---

*[ui/plugins/controller.js, line 387 at r3](https://reviewable.io:443/reviews/mozilla/treeherder/2057#-K__g3iqlVvZniU8vu6n-r3-387:-KatBoj7zTe38ocBDmmt:b-896fix) ([raw file](https://github.com/mozilla/treeherder/blob/9cf1e5c238d2c568700c3ad928b40ebd9c47a3f8/ui/plugins/controller.js#L387)):*
<details><summary><i>Previously, camd wrote…</i></summary><blockquote>

Would you document a bit more how ``buildSignedUrl`` works?  :)
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mozilla/treeherder/2057)*

**Comment by imbstack** [comment]
Review status: 0 of 11 files reviewed at latest revision, 5 unresolved discussions.

---

*[ui/plugins/controller.js, line 394 at r3](https://reviewable.io:443/reviews/mozilla/treeherder/2057#-K__g3iqlVvZniU8vu6n-r3-394:-KatDoXj1t6K_1Xc3Eyi:b-896fix) ([raw file](https://github.com/mozilla/treeherder/blob/9cf1e5c238d2c568700c3ad928b40ebd9c47a3f8/ui/plugins/controller.js#L394)):*
<details><summary><i>Previously, camd wrote…</i></summary><blockquote>

You might consider adding the ``true`` param here to make this error persist.  I don't feel super strongly about this though, so your call.  :)
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mozilla/treeherder/2057)*

**Pull request merged**